### PR TITLE
Deploy: implement .tesselinclude for both --slim and --full path. Closes gh-433

### PR DESCRIPTION
**Pull request by rwaldron**
Not quite ready for full review, I just needed it to hit the appveyor ci.

### CODE REVIEW

File: lib/tessel/deploy.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ var fsTemp = require('fs-temp');
 var browserify = require('browserify');
 var uglify = require('uglify-js');
 var glob = require('glob');
+var rimraf = require('rimraf');
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: TODO: `rimraf` can be removed across the project, in favor of `fs-extra`, which has more capabilities—including an `rm -rf` operation. @rwaldron

File: lib/tessel/deploy.js [hunk 2]
```diff
@@ -308,109 +389,94 @@ actions.tarBundle = function(opts) {
...
-        var rules = ignoreFiles.reduce(function(rules, ignoreFile) {
-          var dirname = path.dirname(ignoreFile);
-          var patterns = fs.readFileSync(ignoreFile, 'utf8').trim().split(/\r?\n/).reduce(function(patterns, pattern) {
-            pattern = pattern.trim();
+      // Inform browserify bundler instance
+      // of all files to exclude. If a file exists in the
+      // inclusion list, then DO NOT exclude it from the
+      // bundling phase.
+      ignoreFiles.forEach(function(file) {
+        if (includeFiles.length &&
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: TODO: unnecessary to check length here @rwaldron

File: lib/tessel/deploy.js [hunk 3]
```diff
@@ -296,9 +297,89 @@ actions.sendBundle = function(tessel, filepath, opts) {
...
+
+      actionsglob.rules(target, '.tesselignore')
+        -> will compile patterns found in all nested .tesselignore files
+
+      actionsglob.rules(target, '.tesselinclude')
+        -> will compile patterns found in all nested .tesselinclude files
+
+   */
+  rules: function(target, filesContainingGlobPatterns) {
+    var files = actions.glob.sync(target + '/**/*' + filesContainingGlobPatterns, {
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: TODO: add comment note about implied path normalization, which is handled by `glob`

File: lib/tessel/deploy.js [hunk 4]
```diff
@@ -308,109 +389,94 @@ actions.tarBundle = function(opts) {
...
+        }
+        bify.exclude(path.join(globRoot, file));
+      });
 
-            // Ignores empty lines and comments
-            if (pattern && !pattern.match(/^#/)) {
-              patterns.push(path.relative(target, path.join(dirname, pattern)));
+      bify
+        .on('file', function() {
+          fileCount++;
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: TODO: add comment note about file counting w/r to its use in optimizing the bundle size

File: lib/tessel/deploy.js [hunk 5]
```diff
@@ -308,109 +389,94 @@ actions.tarBundle = function(opts) {
...
-
-        bify
-          .on('file', function() {
-            fileCount++;
-          })
-          .bundle(function(error, results) {
-            if (error) {
-              return reject(error);
-            } else {
+            var projectDirectory = fsTemp.mkdirSync();
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: TODO: rename `projectDirectory` => `bundleTempDirectory` @rwaldron

File: lib/tessel/deploy.js [hunk 6]
```diff
@@ -308,109 +389,94 @@ actions.tarBundle = function(opts) {
...
-                results = fs.readFileSync(entry);
-              }
+            // Write the primary bundle file
+            fs.writeFileSync(projectBundle, actions.compress(results));
 
-              var projectDirectory = fsTemp.mkdirSync();
-              var projectBundle = path.join(projectDirectory, opts.slimPath);
-
-              fs.writeFileSync(projectBundle, actions.compress(results));
+            // Copy any files that matched all .tesselinclude patterns
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: TODO: can these also be minified? @rwaldron

File: lib/tessel/deploy.js [hunk 7]
```diff
@@ -308,109 +389,94 @@ actions.tarBundle = function(opts) {
...
-        bify
-          .on('file', function() {
-            fileCount++;
-          })
-          .bundle(function(error, results) {
-            if (error) {
-              return reject(error);
-            } else {
+            var projectDirectory = fsTemp.mkdirSync();
+            var projectBundle = path.join(projectDirectory, opts.slimPath);
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: TODO: rename `projectBundle` => `bundleTempFile` @rwaldron

**Comment by HipsterBrown** [comment]
Great work! :+1:

**Comment by rwaldron** [comment]
Here's a simple static file server that demonstrates serving html, favicon.ico, and an image from dirs explicitly specified in https://github.com/rwaldron/tessel-include-bundling

**Pull request merged**